File: cpp/src/actor_config.cpp
// cpp/src/actor_config.cpp
// Purpose: actor_config_from_json — the Port/ACL parse of a `configure` message's "config" object into a
//   validated ActorConfig (ADR-0012 P2 / P9 rule 5: validate-don't-coerce — a typed Error returned by
//   value, never a throw or a coerced default). The field set it reads is the one drift-netted against
//   the Python actor_config.py (tests/test_wire_drift.py); the domain checks mirror schema.check_invariants.
//   nlohmann's accessor exceptions are translated into a typed Error at this boundary (the same
//   discipline instance.cpp uses), so the search/serve core stays throw-free.
//
// Public Domain (The Unlicense).
#include "chocofarm/actor_config.hpp"

#include <string>

namespace chocofarm {

std::expected<ActorConfig, Error> actor_config_from_json(const nlohmann::json& j) {
    if (!j.is_object())
        return std::unexpected(make_error("config must be a JSON object"));
    ActorConfig c;
    try {
        c.instance_path = j.at("instance_path").get<std::string>();
        c.faces_path = j.at("faces_path").get<std::string>();
        // JSON get<int>() is the boundary decode; wrap into the config's typed domains at this ACL (P2).
        c.gumbel.m = CandidateCount{static_cast<SearchRep>(j.at("m").get<int>())};
        c.gumbel.n_sims = SimBudget{static_cast<SearchRep>(j.at("n_sims").get<int>())};
        c.gumbel.c_puct = j.at("c_puct").get<double>();
        c.gumbel.c_visit = j.at("c_visit").get<double>();
        c.gumbel.c_scale = j.at("c_scale").get<double>();
        c.gumbel.c_outcome = OutcomeIndex{static_cast<SearchRep>(j.at("c_outcome").get<int>())};
        c.gumbel.max_depth = PlyDepth{static_cast<SearchRep>(j.at("max_depth").get<int>())};
    } catch (const nlohmann::json::exception& e) {
        // a missing required field (out_of_range) or a wrong type (type_error) — translate into a typed
        // Error at the boundary (P9: the core stays throw-free; the shell reports loudly, ADR-0002).
        return std::unexpected(make_error(std::string("config field error: ") + e.what()));
    }
    // domain validation (validate-don't-coerce; mirrors schema.check_invariants' positive-count + depth
    // domains): a zero/negative budget or an empty geometry path is a typed Error, never a silently-wrong
    // search (an m=0 SH bracket, a missing instance file).
    if (c.instance_path.empty())
        return std::unexpected(make_error("config.instance_path is empty"));
    if (c.faces_path.empty())
        return std::unexpected(make_error("config.faces_path is empty"));
    // domain checks against the typed minima; .value() at the std::to_string diagnostic boundary.
    if (c.gumbel.m < CandidateCount{1})
        return std::unexpected(make_error("config.m must be >= 1, got " + std::to_string(c.gumbel.m.value())));
    if (c.gumbel.n_sims < SimBudget{1})
        return std::unexpected(make_error("config.n_sims must be >= 1, got " + std::to_string(c.gumbel.n_sims.value())));
    if (c.gumbel.c_outcome < OutcomeIndex{1})
        return std::unexpected(make_error("config.c_outcome must be >= 1, got " + std::to_string(c.gumbel.c_outcome.value())));
    if (c.gumbel.max_depth < PlyDepth{1})
        return std::unexpected(make_error("config.max_depth must be >= 1, got " + std::to_string(c.gumbel.max_depth.value())));
    return c;
}

}  // namespace chocofarm

File: chocofarm/az/actor_config.py
#!/usr/bin/env python3
"""
chocofarm/az/actor_config.py — ActorConfig: the C++ Gumbel actor's control config, projected from the
hp-registry ExperimentConfig (ADR-0012 P1/P7 — derive-don't-duplicate across the language boundary).

This is the SSOT for the config knobs the C++ actor consumes: the instance/faces geometry paths and the
7 GumbelConfig search knobs the runner reads as --gumbel-* flags today. It is a DERIVED VIEW of the hp
schema — `from_experiment_config` reads the (already strict-decoded, domain-validated) SearchConfig and
holds NO defaults of its own (the defaults live in schema.py, the one config SSOT). The C++ mirror
`cpp/include/chocofarm/actor_config.hpp` declares the SAME field set + per-field Mut class as parseable
literal arrays, drift-checked against this module in tests/test_wire_drift.py (ADR-0012 P7: the
cross-boundary fact has one authoritative home; every side derives, the drift net is the backstop) —
so a field added/removed/renamed, or a Mut-class flip, on one side reds the default suite rather than
silently desyncing the control config. This module's P1/P7 honesty is CONTINGENT on that drift leg
existing and working (the same status `result_spec` has).

Scope (P2 honesty — only carry what the receiver consumes). ActorConfig carries the knobs the C++
runner ACTUALLY reads. It excludes:
  * use_jax_mlp — a Python-side knob (it selects the JAX vs numpy forward for the in-process eval); the
    C++ runner builds its own NetForward and has no --use-jax-mlp flag, so it never crosses this seam.
  * explore_plies / td_lambda (lam_blend) / n_step — the parity knobs the C++ search cannot honor yet
    (the executor refuses them loudly); they join ActorConfig when the parity thread makes the C++
    search consume them. Putting a field the receiver ignores into the config is the lying signature
    ADR-0002/P2 forbid.

The per-generation scalars (version, seed, lam, episodes, max_steps, res_token) are NOT config — they
ride each generate request (the determinism anchor lives in the per-call message, never sticky config).

Mut classes (read from schema.py's metadata["mut"], the one home — see _SCHEMA_SOURCE): the 7 search
knobs are HOT (the SH bracket is recomputed per decide — they flow live through reconfigure); the two
geometry paths are INSTANCE (built once; a live change is a NEW experiment, a loud reject).

Public Domain (The Unlicense).
"""
from __future__ import annotations

import dataclasses
from dataclasses import dataclass

from chocofarm.hp.schema import EnvConfig, ExperimentConfig, Mut, SearchConfig


@dataclass(frozen=True)
class ActorConfig:
    """The C++ Gumbel actor's control config — instance/faces geometry + the 7 GumbelConfig search
    knobs. A frozen, typed projection of ExperimentConfig (P8: the typed signature is the contract).
    Built via `from_experiment_config`; serialized to the control `configure` message in the transport.
    The fields are FLAT (matching the JSON wire keys); the C++ mirror nests the 7 search knobs in a
    `GumbelConfig` for the runner's use, but the wire field set is flat — see actor_config.hpp."""

    instance_path: str
    faces_path: str
    m: int
    n_sims: int
    c_puct: float
    c_visit: float
    c_scale: float
    c_outcome: int
    max_depth: int


# The field set + each field's home in the hp schema (group, field) — the ONE place ActorConfig's fields
# map back to the SSOT. The Mut class is READ from that schema field's metadata (never declared here), so
# the Mut classification has a single home (the hp schema). instance_path/faces_path both map to
# env.instance_path: faces is the DERIVED geometry of the same instance, so it shares instance's INSTANCE
# class (there is no separate schema field for it — the C++ runner takes the faces path directly).
_SCHEMA_SOURCE: tuple[tuple[str, str, str], ...] = (
    # (ActorConfig field, schema group, schema field)
    ("instance_path", "env", "instance_path"),
    ("faces_path", "env", "instance_path"),
    ("m", "search", "m"),
    ("n_sims", "search", "n_sims"),
    ("c_puct", "search", "c_puct"),
    ("c_visit", "search", "c_visit"),
    ("c_scale", "search", "c_scale"),
    ("c_outcome", "search", "c_outcome"),
    ("max_depth", "search", "max_depth"),
)

# the ordered field names (the drift net compares this against the C++ ACTOR_CONFIG_FIELDS literal).
FIELD_NAMES: tuple[str, ...] = tuple(f for f, _g, _sf in _SCHEMA_SOURCE)

# fail loud at import if the dataclass and the schema-source map disagree (a field added to one and not
# the other is a silent desync the projection must never start under — ADR-0002).
assert FIELD_NAMES == tuple(f.name for f in dataclasses.fields(ActorConfig)), (
    "ActorConfig dataclass fields and _SCHEMA_SOURCE disagree — "
    f"{tuple(f.name for f in dataclasses.fields(ActorConfig))!r} vs {FIELD_NAMES!r}")


_GROUPS: dict[str, type] = {"env": EnvConfig, "search": SearchConfig}


def _mut_of(group: str, field: str) -> Mut:
    """Read the Mut class of one schema (group, field) from its dataclass metadata — the one home. A
    field whose metadata carries no Mut is a loud failure (the schema contract: every leaf has a Mut)."""
    cls = _GROUPS[group]
    for f in dataclasses.fields(cls):
        if f.name == field:
            mut = f.metadata.get("mut")
            assert isinstance(mut, Mut), f"schema {group}.{field} has no Mut facet"
            return mut
    raise KeyError(f"{group}.{field} is not a field of {cls.__name__}")


# the per-field Mut class string ("hot"/"restart"/"instance"), in FIELD_NAMES order — the drift net
# compares this against the C++ ACTOR_CONFIG_MUT literal, so a Mut-class flip on either side reds. The
# strings are exactly the schema Mut enum values (no third vocabulary).
MUT_CLASSES: tuple[str, ...] = tuple(_mut_of(g, sf).value for _f, g, sf in _SCHEMA_SOURCE)


def from_experiment_config(cfg: ExperimentConfig, *, instance_path: str,
                           faces_path: str) -> ActorConfig:
    """Project a (strict-decoded, domain-validated) ExperimentConfig into the actor's ActorConfig.

    Reads the 7 search knobs off `cfg.search` (the hp SSOT — no defaults re-declared here) and takes the
    two geometry paths explicitly (they are the C++ runner's instance/faces, supplied by the executor's
    construction, not an hp schema field — there is no schema home for the C++ faces path). The values
    are already validated by `schema.check_invariants` at decode; this projection adds no new defaults
    and no second validation (Port/ACL: derive, don't re-author)."""
    s = cfg.search
    return ActorConfig(
        instance_path=instance_path,
        faces_path=faces_path,
        m=s.m,
        n_sims=s.n_sims,
        c_puct=s.c_puct,
        c_visit=s.c_visit,
        c_scale=s.c_scale,
        c_outcome=s.c_outcome,
        max_depth=s.max_depth,
    )
